### ISSUE TITLE
Adapters with lower-than-default limits

### ISSUE THREAD

**Issue by eliemichel** [gfx-rs/wgpu-native#297]
From the [3.6.2](https://www.w3.org/TR/webgpu/#limits) section of the spec about limits:

> Every adapter is guaranteed to support the default value or better.

But it turns out that in practice `wgpu-native` may create adapters that have more constrained limits.

For instance, somebody reported to me that they have `max_storage_textures_per_shader_stage` set to 2 in the supported limits, while spec says it must be at least 4. As a consequence, this adapter is not able to create any device using the default limits:

```
Failed to get a WebGPU device: Validation Error

Caused by:
    Limit 'max_storage_textures_per_shader_stage' value 4 is better than allowed 2
```

I see two reasonable ways to address this:

 - **(A)** Stick to WebGPU's spec, and do not create an adapter if it does not support the default limits.
 - **(B)** When device limits are not specified (set to nullptr), use the min of spec's default value and adapter's supported value as the default spec.

Solution **(B)** feels to me like a good trade off, as it does not obsoletes devices that are too limited, but still enables one to ignore the question of limits by setting them to `nullptr` in the device descriptor.

edit: [This quote](https://github.com/gpuweb/gpuweb/issues/1069) confirms that **(A)** isn't necessary thus **(B)** sounds better:

> native, non-web-exposed WebGPU implementations aren't beholden any minimum requirements here, or to the standard. So they may support older or less capable devices with feature emulation or compatibility gaps compared to the official spec.

**Comment by teoxoy** [comment]
> (B) When device limits are not specified (set to nullptr), use the min of spec's default value and adapter's supported value as the default spec.

We don't currently validate required limits (https://github.com/gfx-rs/wgpu/issues/4072) which is required for this to work.

There is also option C: Add another `DownlevelFlags` similar to the `WEBGPU_TEXTURE_FORMAT_SUPPORT` one but for limits.

Related: "strict WebGPU compliance mode" (https://github.com/gfx-rs/wgpu/issues/3274)

**Comment by rajveermalviya** [comment]
We currently use `default()` in following conditions:
- got `NULL` for descriptor in `wgpuAdapterRequestDevice`
- got `NULL` for `WGPUDeviceDescriptor.requiredLimits`
- used as base for limits with value of [`WGPU_LIMIT_[U32/U64]_UNDEFINED`](https://github.com/webgpu-native/webgpu-headers/blob/d12388575785f254b5c186667e0c1b79e4da46b1/webgpu.h#L72-L73) for `WGPUDeviceDescriptor.requiredLimits`

To fix this, we could (and probably should) conditionally use `downlevel_defaults()` as base limits instead of the `default()` limits when adapter's limits are lower than `default()` limits.
But this would happen silently, user will have to get the device limits to know what they received. Though atleast they would get _something_.

Unfortunately we can't do what `wgpu` does, i.e requiring a limits struct and let the user make the decisions, because `webgpu.h` has it optional.

**Comment by rajveermalviya** [comment]
@eliemichel 
I have created a [dummy release](https://github.com/rajveermalviya/wgpu-native/releases/tag/v99.99.99 ) in my fork with the fix backported to v0.17.0.2. Can you ask if it fixes this issue. Thanks!

**Comment by eliemichel** [comment]
Thanks! Will let you know.
